crates/common/logged_command/src/logged_command.rs

```diff
@@ -16,18 +18,60 @@ pub struct LoggingChild {
 
 impl LoggingChild {
     pub async fn wait_with_output(
-        self,
+        mut self,
         logger: &mut BufWriter<File>,
+        time_out: Option<usize>,
     ) -> Result<Output, std::io::Error> {
+        match time_out {
+            Some(timeout) => {
+                tokio::select! {
+                    _ = self.inner_child.wait() => {
+                        self.get_outcome(logger).await
+                    }
```

Comment:
Something is weird here as we start to await the child with `self.inner_child.wait()` *ignoring* its output, then when terminated we await again, using that time `self.inner_child.wait_with_output()`. Reading the [documentation](https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.wait_with_output) it's not clear which *remaining* output is collected.

Why not something along these lines:
```suggestion
                    outcome = self.inner_child.wait_with_output() => {
                        self.finalize_outcome(logger, outcome).await
                    }
```

Reply:
If we do as suggested above I doubt we cover all the below cases
```
match time_out {
 **// If the timeout is provided then get  inside**       
         Some(timeout) => { 
                tokio::select! {
          
         **// if the operation completes before timeout then get the outcome**   
                    _ = self.inner_child.wait() => {                      self.get_outcome(logger).await
                    }
                   **// if the operation did not complete within the given time, then kill it forcefully, collect the outcome, and update the error message accordingly.**
                    _ = tokio::time::sleep(timeout) => {
                        self.timeout_operation(logger).await
                    }
                }
            }
            **// If the timeout is not provided then, wait till it completes then collect the outcome**
            None => self.get_outcome(logger).await,
        }
```

Reply:
The whole `tokio::select!` logic has to be rewritten and tests added to cover the 4 cases (no timeout, the script terminates before the timeout, the script is killed with SIGTERM, the script is killed with SIGKILL).

Reply:
Added tests to test all the four cases

Reply:
> > Added tests to test all the four cases
> 
> Thank you. However, these tests doesn't cover the point I would like to clarify and the issue I see to start with `inner_child.wait()` and use `inner_child.wait_with_output()` only _after_ a timeout. **Is the output of the command correctly collected?** 

Yes, we are collecting the output correctly. The `inner_child.wait()`, just returns the exit status, when it finishes successfully. The outcome of the `command` (success or killed) has to be collected using the `inner_child.wait_with_output()`. This is what we are doing here.

> The scripts used for these tests must echo some content and the tests must check the content emitted before the script has been killed is as expected.

Yes, I did echo some text for only one test and then asserted on that content as well. I can do the same for all the other tests as well.

> 
> ```
>     let script = r#"#!/usr/bin/bash    
>     for i in {1..50}
>     do
>         echo $i
>         sleep 1
>     done
>     "#;
> ```
> 
> As side note, it might be simpler to write these tests here - at the level of the `logged_command`, rather than behind the mapper.

I thought about it. Then decided to go at the `mapper` level because there were no tests for the `custom` operations at the mapper level. This will cover now both the `logged_command` as well as the mapper level.

Reply:
Updated the tests to assert the `execution status and stdout/stderr ouptut` that is logged.

Reply:
> Yes, we are collecting the output correctly. 

I'm still not convinced by the tests as only the latest line is checked.

Reply:
Do you want to assert the whole text of the log?

Reply:
The improved version of  `timeout_operation()` is far better now with a second `tokio::select!` to await either a timeout or the child to exit gracefully. However, I would really prefer that a *single* future is created to await the child and to *re-use* the same future in all the timeout checks. The tests are still only not checking the whole content of the child stdout and stderr while it's really not clear what part of these logs will be actually captured when calling `inner_child.wait_with_output()` only after the termination of the child.

Reply:
I tried to have a single future, but there is an issue with sharing the child IDs across two different futures. 
The tests are now asserting the whole output of the logs. 

The inner_child_with_output, will capture the entire output of the command execution, it includes the stdout, stderr, and also the exit status.

Reply:
My comment from April 21st is still valid. I don't understand how this can work.

Sorry, but I find the current proposal contrived and not convincing. I cannot blindly approved it.
So, I will take the time to understand how the whole output of the child can be captured while being ignored at first.